### PR TITLE
Vaults pausing

### PR DESCRIPTION
**Pull request by crisgarner**

### CODE REVIEW

**Review by voith — COMMENTED** [review]
I haven't read the tests yet. But I've left some feedback on the code changes. 
Besides the code review comments, here's a gas golfing idea. 
Instead of using a map for `isDisabled` we could use an 8bit integer and use bitwise logic to check if a bit at a particular position is 1 or 0. 

Here's the python equivalent of the code I have in mind:
```python
from enum import Enum

class DisableFunctionChoices(Enum):
	CreateVault = 0
	AddCollateral = 1
	RemoveCollateral = 2
	Mint = 3
	Burn = 4
	LiquidateVault = 5

isDisabled = 0x00000000;

def disableFunction(_function: int) -> None:
	global isDisabled  # specific to python
	bit_mask = (1 << _function)
	assert ((isDisabled & bit_mask) != bit_mask), "function already disabled"
	isDisabled = (isDisabled | bit_mask)

def enableFunction(_function: int) -> None:
	global isDisabled  # specific to python
	bit_mask = (1 << _function)
	assert ((isDisabled & bit_mask) == bit_mask), "function already enabled"
	isDisabled = (isDisabled & ~bit_mask)

def whenNotDisabled(_function: int) -> bool:
	global isDisabled  # specific to python
	bit_mask = (1 << _function)
	# change return to require in solidity, to use this as a modifier
	return ((isDisabled & bit_mask) != bit_mask)
```

some tests:
```
In [62]: isDisabled
Out[62]: 0

In [63]: disableFunction(DisableFunctionChoices.CreateVault.value)

In [64]: disableFunction(DisableFunctionChoices.LiquidateVault.value)

In [65]: whenNotDisabled(DisableFunctionChoices.LiquidateVault.value)
Out[65]: False

In [66]: whenNotDisabled(DisableFunctionChoices.AddCollateral.value)
Out[66]: True

In [67]: whenNotDisabled(DisableFunctionChoices.CreateVault.value)
Out[67]: False

In [68]: whenNotDisabled(DisableFunctionChoices.Mint.value)
Out[68]: True

In [69]: enableFunction(DisableFunctionChoices.CreateVault.value)

In [70]: bin(isDisabled)
Out[70]: '0b100000'

In [71]: enableFunction(DisableFunctionChoices.CreateVault.value)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-71-4768a214e741> in <module>()
----> 1 enableFunction(DisableFunctionChoices.CreateVault.value)

<ipython-input-61-ef204ef8a55b> in enableFunction(_function)
     20         global isDisabled
     21         bit_mask = (1 << _function)
---> 22         assert ((isDisabled & bit_mask) == bit_mask), "function already enabled"
     23         isDisabled = (isDisabled & ~bit_mask)
     24

AssertionError: function already enabled

In [72]: whenNotDisabled(DisableFunctionChoices.CreateVault.value)
Out[72]: True

In [73]: whenNotDisabled(DisableFunctionChoices.LiquidateVault.value)
Out[73]: False

In [74]:  disableFunction(DisableFunctionChoices.Mint.value)

In [75]: bin(isDisabled)
Out[75]: '0b101000'

In [76]: whenNotDisabled(DisableFunctionChoices.Mint.value)
Out[76]: False

In [77]: whenNotDisabled(DisableFunctionChoices.CreateVault.value)
Out[77]: True
```

**Comment by voith** [comment]
I'll review the tests once you address my comments.

File: contracts/IVaultHandler.sol [hunk 1]
```diff
@@ -181,6 +186,13 @@ IERC165
 	/// @notice An event emitted when a erc20 token is recovered
 	event Recovered(address _token, uint256 _amount);
 
+	/// @notice An event emitted when a function is paused
+    event FunctionIsPaused(
```
[the remark this answers is not in the document]
Comment (crisgarner) [reply contracts/IVaultHandler.sol:190]: Remade to disabled

File: contracts/IVaultHandler.sol [hunk 2]
```diff
@@ -347,12 +359,18 @@ IERC165
 		emit NewTreasury(msg.sender, _treasury);
 	}
 
+    function togglePauseFunction(uint8 _function, bool _isPaused) external virtual onlyOwner {
+        isPaused[_function] = _isPaused;
+        emit FunctionIsPaused(msg.sender, _function,_isPaused);
```
[the remark this answers is not in the document]
Comment (crisgarner) [reply contracts/IVaultHandler.sol:364]: Renamed the event  and function:

```
event FunctionToggled(
    address indexed _owner,
    uint256 _function,
    bool _isDisabled
  );

function toggleFunction(uint8 _function, bool _isDisabled)
```

File: contracts/IVaultHandler.sol [hunk 3]
```diff
@@ -93,7 +102,11 @@ IERC165
 	/// @notice Id To Vault
 	mapping(uint256 => Vault) public vaults;
 
-	/// @notice value used to multiply chainlink oracle for handling decimals
+    /// @notice mapping of function to is Paused
+    /// @dev 1 = CreateVault, 2 = AddCollateral, 3 = RemoveCollateral, 4 = Mint, 5 = Burn, 6 = LiquidateVault
```
Comment (voith) [inline contracts/IVaultHandler.sol:106]: Now that we are using enums, this comment is no longer needed.

File: contracts/IVaultHandler.sol [hunk 4]
```diff
@@ -93,7 +102,11 @@ IERC165
 	/// @notice Id To Vault
 	mapping(uint256 => Vault) public vaults;
 
-	/// @notice value used to multiply chainlink oracle for handling decimals
+    /// @notice mapping of function to is Paused
```
Comment (voith) [inline contracts/IVaultHandler.sol:105]:
```suggestion
    /// @notice map that holds information about a function being enabled/disabled.
```

File: contracts/IVaultHandler.sol (hunk 3 above)
Comment (voith) [inline contracts/IVaultHandler.sol:106]: Now that we are using enums, this comment is no longer needed.

File: contracts/IVaultHandler.sol [hunk 5]
```diff
@@ -181,6 +194,13 @@ IERC165
 	/// @notice An event emitted when a erc20 token is recovered
 	event Recovered(address _token, uint256 _amount);
 
+	/// @notice An event emitted when a function is disabled
```
Comment (voith) [inline contracts/IVaultHandler.sol:197]:
```suggestion
	/// @notice An event that is emitted when a function is enabled/disabled
```

File: test/OptimisticTreasury.t.sol [hunk 6]
```diff
@@ -67,8 +64,9 @@ contract OptimisticTreasuryTest is DSTest {
 		}
 	}
 
-	function testRetrieveEth(address _to) public {
-		if(_to == address(0x01) || address(this) == _to) return;
+	function testRetrieveEth() public {
+        address _to = address(0x11);
```
Comment (voith) [inline test/OptimisticTreasury.t.sol:68]: Sorry, I don't understand this. Why is this line still there?
Line 69 is useless when you hardcode this address.

File: test/VaultsPausing.t.sol [hunk 7]
```diff
@@ -0,0 +1,496 @@
...
+    assertEq(ethVault.userToVault(user), 0);
+  }
+
+  function testCreateVault_ShouldWork_WhenToogleDisabledFalse() public {
+    //setUp
+    vm.prank(address(orchestrator));
+    ethVault.toggleFunction(IVaultHandler.FunctionChoices.CreateVault, true);
+    vm.prank(user);
+    vm.expectRevert("VaultHandler:: function disabled");
+    ethVault.createVault();
```
Comment (voith) [inline test/VaultsPausing.t.sol:105]: Also test that UserToVault was not set:
```solidity
  assertEq(ethVault.userToVault(user), 0);
```

**Review by voith — COMMENTED** [review]
Besides the minor requests, I like to see one more test. 
Say if we start with two **disabled** functions, an then enable of the previously disabled functions then I want to be sure that the disabled function still remains disabled:

```solidity
function test(){
  ethVault.toggleFunction(IVaultHandler.FunctionChoices.CreateVault, true)
  ethVault.toggleFunction(IVaultHandler.FunctionChoices.AddCollateral, true);
  assertEq(ethVault.isDisabled(IVaultHandler.FunctionChoices.CreateVault), true);
  assertEq(ethVault.isDisabled(IVaultHandler.FunctionChoices.AddCollateral), true);

  ethVault.toggleFunction(IVaultHandler.FunctionChoices.CreateVault, false);
  assertEq(ethVault.isDisabled(IVaultHandler.FunctionChoices.CreateVault), false);
  assertEq(ethVault.isDisabled(IVaultHandler.FunctionChoices.AddCollateral), true);
}
```

**Comment by voith** [comment]
@crisgarner Delete `lib/forge-std` directory. It's not needed. I removed it in #122, but you added it again in this PR. CI will pass after you do that.

File: test/OptimisticTreasury.t.sol (hunk 6 above)
Comment (crisgarner) [reply test/OptimisticTreasury.t.sol:68]: yes I agree, this was fixed on a past commit, I haven't published yet

File: .gitmodules [hunk 8]
```diff
@@ -4,3 +4,9 @@
 [submodule "lib/openzeppelin-contracts"]
 	path = lib/openzeppelin-contracts
 	url = https://github.com/OpenZeppelin/openzeppelin-contracts
+[submodule "forge-std"]
```
Comment (voith) [inline .gitmodules:7]: What's the reason behind adding this? We already have `forge-std` in `lib/forge-std`. Why did you add another one to the base directory?

Comment (crisgarner) [reply .gitmodules:7]: This is a mistake, has been removed now.

**Review by voith — APPROVED** [review]
LGTM 👍 

:shipit:
